include/dem/adaptive_sparse_contacts.h

```diff
@@ -0,0 +1,522 @@
+   */
+  void
+  calculate_granular_temperature_and_solid_fraction(
+    const Particles::ParticleHandler<dim> &particle_handler,
+    const std::set<typename DoFHandler<dim>::active_cell_iterator>
+                   &local_and_ghost_cells_with_particles,
+    Vector<double> &cell_granular_temperature,
+    Vector<double> &cell_solid_fraction);
+
+  /**
+   * @brief Assign the mobility status to the cell, its nodes, and its periodic
+   * coinciding nodes from the periodic nodes map. If required, this will
+   * prevent overwriting a node status prevailing over the initial one.
+   * For instance, empty status must not be overwritten by mobile status because
+   * all cells with this node will be part of the additional mobile layer and
+   * will wrongly propagate mobility status to the next cells.
+   *
+   * @param cell_id The current cell index
+   *
+   * @param dof_indices The vector of the DoF indices of the cell
+   *
+   * @param node_id The current node index (periodic or not)
+   *
+   * @param mobility_status The mobility status to assign
+   *
+   */
+  inline void
+  assign_mobility_status(unsigned int                          cell_id,
+                         std::vector<types::global_dof_index> &dof_indices,
+                         const int                             cell_status,
+                         const int                             node_status)
+  {
+    cell_mobility_status.insert({cell_id, cell_status});
+
+    // Assign mobility status to nodes without overwriting empty or mobile nodes
+    // in regard to the case
+    for (auto node_id : dof_indices)
+      {
+        // Prevailing mobility status of the node and assignation
+        int status_assigned = std::max(node_status, mobility_at_nodes(node_id));
+        mobility_at_nodes(node_id) = status_assigned;
+
+        // Check if node has a periodic node and assign the same mobility status
+        // if prevailing on the one on the periodic node
+        auto it = periodic_node_ids.find(node_id);
+        if (it != periodic_node_ids.end())
+          {
+            mobility_at_nodes(it->second) =
+              std::max(status_assigned, mobility_at_nodes(it->second));
+          }
+      }
+  }
+
+  /**
+   * @brief Assign the mobility status to the cell. No need to update the node
+   * status. Only use for active status assignment since nodes are not evaluated
+   * after this step.
+   *
+   * @param cell_id The current cell index
+   *
+   * @param mobility_status The mobility status to assign
+   *
+   */
+  inline void
+  assign_mobility_status(unsigned int cell_id, const int cell_status)
+  {
+    cell_mobility_status.insert({cell_id, cell_status});
+  }
+
+  // Set of locally owned and ghost cells: <local/ghost cells>
+  // Used to loop over only the locally owned and ghost cells without looping
+  // over all the cells in the triangulation numerous times.
+  std::set<typename DoFHandler<dim>::active_cell_iterator>
```

Comment:
would an std::unordered be faster here?
I think in general if a hash function exist hash sets are faster than regular sets

Reply:
There is no hash for cell iterator